### PR TITLE
feat: Add send notification task to locust load test file

### PR DESCRIPTION
**Pull request by jrbenny35**
This adds a send notification task to locust. It also updates the register task to handle errors a bit better.

[SYNC-3846](https://mozilla-hub.atlassian.net/jira/software/c/projects/SYNC/issues/SYNC-3846)

Note: I didn't add the unregister function as it seems like we don't need that currently, but it can be added if needed.

[SYNC-3846]: https://mozilla-hub.atlassian.net/browse/SYNC-3846?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by pjenvey** [comment]
Rebasing against master will fix the ci failures

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:182]: This while loop can be removed.

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:191]: This is also doing flow-of-control through a try-catch, which should be avoided.

In the case there is an error instead of a retry, we should report the error in the performance metrics.
It could be we need to stop the user after.

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:157]: In the case where we are re-connecting. I think we want to re-use the `uaid`? I think we need to add support in the hello() method for this.

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:213]: returning is preferable to raising ResheduleTask()

At the moment how do we end up in a state where `channels` is Falsy? We register a channel on start and stop the user if it doesn't succeed. We also don't have an unregister method yet. Do we need this check at this point in time?

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:217]: We shouldn't use the `_time_event` context when using the `client`. Instead use a with statement with the `client` and I think 'catch_response' must be set to True.

```
with self.client.post(..., catch_response=True, ... ) as response:
```

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:225]: Because we are using the `self.client` an assertion isn't the appropriate choice for verification. Instead we change the response status. Something like:

```suggestion
            if response.status_code != 201:
                response.failure(f"{response.status_code=}, expected 201, {response.text=}")
                return
```

**Review by Trinaa — CHANGES_REQUESTED** [review]
Sorry @jrbenny35 I'm asking for a lot of changes. If it's easier we can go through them together in pair.

Also, have you had a chance to run the test in GCP?
Before merging this change we'll need to evaluate the impact and adjust the Users and Spawn Rate if needed (ref: [Previous Results](https://docs.google.com/spreadsheets/d/1-nF8zR98IiLZlwP0ISwOB8cWRs74tJGM1PpgDSWdayE/edit#gid=0)) Let me know if you'd like to do that together.

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:157]: Check if user already has UAID and reuse it

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:161]: Check Status code and response length

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:164]: Remove try/catch and try and receive the notification

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:182]: Should bail is failure

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:217]: Client already supports time tracking

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:181]:
```suggestion
        with self._time_event(name="register") as timer:
```

**Comment by Trinaa** [comment]
@jrbenny35 I noticed that 'hello' can have 'channelIDs' as part of it's payload. Perhaps the absence of this information on re-connection is causing some of the odd behaviour.

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [reply tests/load/locustfiles/locustfile.py:161]: Further investigation reveals that the response can be modelled as follows:

```
class AckMessage(BaseModel):
    """Autopush 'ack' response message."""
    data: str
    headers: dict[str, str]
    messageType: Literal["notification"]
    channelID: str
    version: str
```

There is no status in this case, so I suggest that we verify using a combination of 'channelID' and 'data'. They should match the values that were posted.

File: tests/load/locustfiles/locustfile.py
Comment (pjenvey) [inline tests/load/locustfiles/locustfile.py:153]: Only hello needs `use_webpush`
```suggestion
                dict(messageType="ack", updates=dict(channelID=channel_id))
```

**Review by pjenvey — CHANGES_REQUESTED** [review]
Also I noticed we're passing `ssl=False` to `create_connection`. The websocket connection definitely forces SSL so I believe this is an invalid argument that the library's ignoring (closest I see is its `sslopt` argument)

File: tests/load/locustfiles/locustfile.py
Comment (pjenvey) [inline tests/load/locustfiles/locustfile.py:158]: The register failing is due to the `ack` being malformed, causing autopush to not be in a happy state when the register was issued:
```suggestion
                    messageType="ack",
                    updates=[dict(channelID=notification.channelID, version=notification.version)]
```

File: tests/load/locustfiles/locustfile.py
Comment (pjenvey) [inline tests/load/locustfiles/locustfile.py:169]: Why do we sleep here, is it necessary?

**Comment by pjenvey** [comment]
Here's a [branch](https://github.com/mozilla-services/autopush-rs/compare/sync-3846-2...sync-3846-2-direct-param) that makes the disconnect/reconnect optional via a parameter. The disconnect causes the notification to be written to DynamoDB: I think it's useful to have both options.

A mixture of "both" would be nice to have -- ideally via task weighting -- but I'm not quite sure how a parameter could easily drive that.

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:129]: Something is not evaluating correctly here, so that i'm not able to execute the load tests. Error message:
image

I've kept the default input:
image

File: tests/load/locustfiles/models.py
Comment (Trinaa) [inline tests/load/locustfiles/models.py:27]: Instead of an assignment, I think we probably want a literal. Should the register message return without a status an assignment would add a 200, which would give us a false positive.

```suggestion
    status: Literal[200]
```

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:248]: suggestion

```suggestion
        channel_id, endpoint_url = random.choice(list(self.channels.items()))

```

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:238]: nit
```suggestion
        Sends a notification to a registered endpoint
```

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:250]: I think this is a left over from a merge. 
```suggestion
```

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:265]: In the case where we send 'stored' messages, a failure to post would result in the user becoming disconnected because we return on line 260.

I solved this locally by extracting the post logic into another method, and checking for success before calling self.ack().

```suggestion
        notification_success: bool = self.post_notification(endpoint_url)

        if stored:
            self.connect()
            self.hello()

        if not notification_success:
            return

        self.ack()
```

```
    def post_notification(self, endpoint_url) -> bool:
        """
        Send a notification to Push

        Args:
            endpoint_url: A channel destination endpoint url
        Returns:
            bool: Flag indicating if the post notification request was successful
        """
        with self.client.post(
                url=endpoint_url,
                name="notification",
                data=self.encrypted_data,
                headers=self.headers,
        ) as response:
            if response.status_code != 201:
                response.failure(f"{response.status_code=}, expected 201, {response.text=}")
                return False
        return True
```

**Review by Trinaa — CHANGES_REQUESTED** [review]

File: tests/load/locustfiles/models.py
Comment (jrbenny35) [reply tests/load/locustfiles/models.py:27]: Ah, I didn't mean to include that. Will remove it

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:265]: Could we just ```else```  and not return? 
```suggestion
        with self.client.post(
            url=endpoint_url,
            name="Endpoint Notification",
            data=self.encrypted_data,
            headers=self.headers,
        ) as response:
            if response.status_code == 201:
                 if stored:
                     self.connect()
                     self.hello()
                 self.ack()
             else:   
                 response.failure(f"{response.status_code=}, expected 201, {response.text=}")
```

Comment (Trinaa) [reply tests/load/locustfiles/locustfile.py:265]: I think for the else to work it would have to look like this right 😬?
```
if response.status_code == 201:
    if stored:
        self.connect()
        self.hello()
    self.ack()
else:
    response.failure(f"{response.status_code=}, expected 201, {response.text=}")
    if stored:
        self.connect()
        self.hello()
```

Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:265]: Oh so we need to run ```connect``` and ```hello``` even on a failure? If the endpoint doesn't return a 201 do we need to reconnect or should we just hope the user (through locust) tries to send another notification and when that passes we reconnect and send the ```ack```

Comment (Trinaa) [reply tests/load/locustfiles/locustfile.py:265]: Leaving the users connected is part of the load we are trying to re-create is my understanding and in future when we have more tasks, what if the next task chosen isn't 'send_notification', we would be leaving the user in an unexpected state.

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [reply tests/load/locustfiles/locustfile.py]: This is resolved in https://github.com/mozilla-services/autopush-rs/pull/438

**Comment by Trinaa** [comment]
Hi @jrbenny35 

Changes from our pair session today --> https://github.com/mozilla-services/autopush-rs/pull/438

**Comment by jrbenny35** [comment]
Merged in @Trinaa's branch. https://github.com/mozilla-services/autopush-rs/pull/438/files

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [inline tests/load/locustfiles/locustfile.py:177]: This can be removed, since the HelloMessage model now enforces it.
```suggestion
```

**Review by Trinaa — COMMENTED** [review]
There has been a lot of churn on this PR and I thank you @jrbenny35 for coordinating all the debugging that was going on. I'm going to suggest we clean-up a bit and stop here for the scope of [SYNC-3846](https://mozilla-hub.atlassian.net/browse/SYNC-3846). We can continue with the following tasks:

* [SYNC-3914](https://mozilla-hub.atlassian.net/browse/SYNC-3914) - This bug was discovered during the implementation of this task and the solution looks extensive enough to warrant an individual ticket and PR.
* [SYNC-3915](https://mozilla-hub.atlassian.net/browse/SYNC-3915) - This work was originally scoped for this PR, but is blocked by [SYNC-3914](https://mozilla-hub.atlassian.net/browse/SYNC-3914), so I think it's best to continue after the bug is resolved.
* [SYNC-3916](https://mozilla-hub.atlassian.net/browse/SYNC-3916) - As @pjenvey suggested, I think we should add this functionality in a follow-up epic.

What do you think?

PS: Should we squash these commits before merging?

**Comment by jrbenny35** [comment]
Bit of a scary squash but I think I got it

File: tests/load/locustfiles/locustfile.py
Comment (jrbenny35) [reply tests/load/locustfiles/locustfile.py:189]: Ah. that's what I get for not running it before a push

**Review by Trinaa — DISMISSED** [review]
🚀 Thanks for the hard work @jrbenny35!

File: tests/load/locustfiles/locustfile.py
Comment (pjenvey) [inline tests/load/locustfiles/locustfile.py:243]:
```suggestion
            assert response, "No 'register' response"
```

File: tests/load/locustfiles/locustfile.py
Comment (pjenvey) [inline tests/load/locustfiles/locustfile.py:137]: I wonder if this is too high -- it might be preferable to fail faster when things aren't working as expected.

**Review by pjenvey — CHANGES_REQUESTED** [review]

File: tests/load/locustfiles/locustfile.py
Comment (Trinaa) [reply tests/load/locustfiles/locustfile.py:137]: oh! that's my bad, at one point I thought the timeout had to be greater than our wait_time interval. What's a good amount 5?

Comment (pjenvey) [reply tests/load/locustfiles/locustfile.py:137]: It's very short in our integration tests (1) but 5 works

**Review by pjenvey — APPROVED** [review]

**Review by Trinaa — APPROVED** [review]

**Pull request merged**
